### ISSUE TITLE
Entity Rigger Component not implementing geometry provider

### ISSUE THREAD

**Issue by peterlindgren** [OurMachinery/themachinery-public#676]
I noticed that if I try to cook a physics mesh (instead of using the box shape type) in the box.entity it fails with the following error:

> Couldn't extract geometry from entity `box`. No DCC Asset component or other geometry providing component found.

So I suppose it's the Entity Rigger Component that's not implemented as a geometry provider?

The reason I was attempting to cook the physics mesh is because I'm working on a similar plugin which also cooks geometry and I'm having problems retrieving the DCC Asset geometry when the entity uses the Entity Rigger Component.

In prior versions (mid 2020) the entity rigger component header exposed the truth properties which I could use to retrieve the DCC Asset:

```#define TM_TT_TYPE__ENTITY_RIGGER__COMPONENT "tm_entity_rigger_component"
#define TM_TT_TYPE_HASH__ENTITY_RIGGER__COMPONENT TM_STATIC_HASH("tm_entity_rigger_component", 0xb64dd8368c303ffdULL)

enum {
    TM_TT_PROP__ENTITY_RIGGER_COMPONENT__DCC_ASSET, // reference(TM_TT_TYPE__ASSET)
    TM_TT_PROP__ENTITY_RIGGER_COMPONENT__MESH_CREATION_GRAPH // reference(TM_TT_TYPE__ASSET)
};

**Comment by niklas-ourmachinery** [comment]
@tobias-ourmachinery Any ideas about this one?

**Comment by peterlindgren** [comment]
I forgot to mention that by setting up the same defines and enum values I was able to reproduce the old behavior, like this:

`#define TM_TT_TYPE_HASH__ENTITY_RIGGER_COMPONENT TM_STATIC_HASH("tm_entity_rigger_component", 0xb64dd8368c303ffdULL)

enum {
    TM_TT_PROP__ENTITY_RIGGER_COMPONENT__DCC_ASSET, // reference(TM_TT_TYPE__ASSET)
    TM_TT_PROP__ENTITY_RIGGER_COMPONENT__MESH_CREATION_GRAPH // reference(TM_TT_TYPE__ASSET)
};

...

static tm_tt_id_t tt_get_dcc_asset(const tm_the_truth_o *tt, const tm_tt_id_t entity_asset) {
    const tm_tt_type_t dcc_asset_type = tm_the_truth_api->object_type_from_name_hash(tt, TM_TT_TYPE_HASH__DCC_ASSET_COMPONENT);
    const tm_tt_id_t dcc_asset_object = tm_the_truth_api->find_subobject_of_type(tt, tm_tt_read(tt, entity_asset), TM_TT_PROP__ENTITY__COMPONENTS, dcc_asset_type);
    if (dcc_asset_object.u64) {
        return tm_the_truth_api->get_reference(tt, tm_tt_read(tt, dcc_asset_object), TM_TT_PROP__DCC_ASSET_COMPONENT__DCC_ASSET);
    }

        const tm_tt_type_t entity_rigger_type = tm_the_truth_api->object_type_from_name_hash(tt, TM_TT_TYPE_HASH__ENTITY_RIGGER_COMPONENT);
        const tm_tt_id_t entity_rigger_object = tm_the_truth_api->find_subobject_of_type(tt, tm_tt_read(tt, entity_asset), TM_TT_PROP__ENTITY__COMPONENTS, entity_rigger_type);
        if (entity_rigger_object.u64) {
            return tm_the_truth_api->get_reference(tt, tm_tt_read(tt, entity_rigger_object), TM_TT_PROP__ENTITY_RIGGER_COMPONENT__DCC_ASSET);
        }

    return (tm_tt_id_t){0};
}`

This is probably a terrible idea. I also haven't worked on this code for some time now so YMMV...

**Comment by tobias-ourmachinery** [comment]
I think you are seeing two different problems here:

1) One problem is that we are currently lacking a way to cook a physics mesh from a `GPU Geometry`. We can fix that by reading back the vertex stream for positions and index buffer to CPU and pass that to the physics cooker. I've added a new issue for this that I'll try to get done for our October release (https://github.com/OurMachinery/themachinery-public/issues/816). This would solve the error message about not being able to cook physics for the box entity. 

2) The other issue appears to be that you are working on old data that uses the `dcc_asset_component` and the `entity_rigger_component`. The `dcc_asset_component` has been deprecated and the entity rigger has been removed completely in later versions of the engine. As you've discovered you can still access the data from those as we don't remove any data from the truth that we don't understand. I think the appropriate way for automatically adding physics shapes during import of new assets would be to add that to the `tm_entity_rigger_api`. Or if you setup your own plugin that adds the data after the entity has been rigged. 

Feel free to reach out to me and @karl-zylinski-ourmachinery to discuss this in more detail as we'd like to better understand what you want to do and help come up with a clean way doing it. 

Closing this issue for now.

**Comment by peterlindgren** [comment]
Hi Tobias,

Thanks for the great info. I'd like to continue the discussion in here, if you don't mind? I find Discord is difficult to keep track of unless I'm actively using it a lot :)

> As you've discovered you can still access the data from those as we don't remove any data from the truth that we don't understand.
I suspected this when I was looking at the data on disk but wasn't sure. That leads me to the question, are there any tools for scrubbing deprecated data?

> Or if you setup your own plugin that adds the data after the entity has been rigged.
I wanted to create a cooking feature, similar to the Physics Shape Component, but in my case it's a "Navmesh Shape" component for which I want to cook the graphics (or physics) mesh data into a navigation mesh representation. I have source access now and could probably copy what Physics Shape component does for the graphical mesh data.

I hope that makes it a bit clearer as to what I was trying to accomplish when I ran into this.

-Peter
